8263384: IGV: Outline should highlight the Graph that has focus

robcasloz opened the pull request:
This changeset eases navigation within and across graph groups by highlighting the focused graph in the Outline window. If the user changes the focus to another graph window, or moves to the previous or next graph within the same window, the newly focused graph is automatically highlighted in the Outline window. This is implemented by maintaining a static map from opened graphs to their corresponding [NetBeans nodes](https://netbeans.apache.org/tutorials/nbm-selection-2.html). The Outline window uses the map to select, on a graph focus change, the NetBeans node of the newly focused graph that should be highlighted.

Tested manually by opening simultaneously tens of graphs from different groups and switching the focus randomly.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8263384](https://bugs.openjdk.org/browse/JDK-8263384): IGV: Outline should highlight the Graph that has focus


### Reviewers
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer)
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9167/head:pull/9167` \
`$ git checkout pull/9167`

Update a local copy of the PR: \
`$ git checkout pull/9167` \
`$ git pull https://git.openjdk.org/jdk pull/9167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9167`

View PR using the GUI difftool: \
`$ git pr show -t 9167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9167.diff">https://git.openjdk.org/jdk/pull/9167.diff</a>

</details>

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:
```diff
@@ -205,6 +217,19 @@ protected boolean requestFocusInWindow(boolean temporary) {
     @Override
     public void resultChanged(LookupEvent lookupEvent) {
+        // Highlight the focused graph, if available, in the outline.
+        if (result.allItems().isEmpty()) {
```
It looks like you are confident that result is not NULL when resultChanged() is called.  I believe framework calls it after `componentOpened()`.  if so, is it possible to remove result = null in `componentClosed()`?

robcasloz replied:
Right, see e.g. the code listed in https://netbeans.apache.org/tutorials/nbm-selection-1.html#_creating_a_context_sensitive_topcomponent which follows the same pattern.

`result = null` is just added for consistency with other TopComponent classes in the project, e.g:

https://github.com/openjdk/jdk/blob/b2a58bec4a4f70a06b23013cc4c351b36a413521/src/utils/IdealGraphVisualizer/ControlFlow/src/main/java/com/sun/hotspot/igv/controlflow/ControlFlowTopComponent.java#L134

But you are right, it does not seem necessary and I just removed it.

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:
```diff
@@ -205,6 +217,19 @@ protected boolean requestFocusInWindow(boolean temporary) {
     @Override
     public void resultChanged(LookupEvent lookupEvent) {
+        // Highlight the focused graph, if available, in the outline.
+        if (result.allItems().isEmpty()) {
```
It looks like you are confident that result is not NULL when resultChanged() is called.  I believe framework calls it after `componentOpened()`.  if so, is it possible to remove result = null in `componentClosed()`?

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:
```diff
@@ -205,6 +217,19 @@ protected boolean requestFocusInWindow(boolean temporary) {
+            return;
+        }
+        try {
+            manager.setSelectedNodes(new GraphNode[]{FolderNode.getGraphNode(p.getGraph())});
```
Do we need to consider that FolderNode.getGraphNode() returns null? `setSelectedNodes` will throw `IllegalArgumentException` if input is null?

robcasloz replied:
`FolderNode.getGraphNode()` should never return null, so if that happens I think the best option is to warn the user and log the `IllegalArgumentException` exception thrown by `setSelectedNodes()`, which is what `Exceptions.printStackTrace()` does. See https://bits.netbeans.org/12.6/javadoc/org-openide-explorer/org/openide/explorer/ExplorerManager.html#setSelectedNodes-org.openide.nodes.Node:A- and https://bits.netbeans.org/12.6/javadoc/org-openide-util/org/openide/util/Exceptions.html#printStackTrace-java.lang.Throwable-.

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/OutlineTopComponent.java:
```diff
@@ -205,6 +217,19 @@ protected boolean requestFocusInWindow(boolean temporary) {
+            return;
+        }
+        try {
+            manager.setSelectedNodes(new GraphNode[]{FolderNode.getGraphNode(p.getGraph())});
```
Do we need to consider that FolderNode.getGraphNode() returns null? `setSelectedNodes` will throw `IllegalArgumentException` if input is null?

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/FolderNode.java:
```diff
@@ -56,15 +61,26 @@ public FolderChildren(Folder folder) {
+        protected void destroyNodes(Node[] nodes) {
+            for (Node n : nodes) {
+                // Each node is only present once in the graphNode map.
+                graphNode.values().remove(n);
```
Is `destroyNodes()` thread-safe here? graphNode is a HashMap instead of ConcurrentHashMap.

navyxliu commented on src/utils/IdealGraphVisualizer/Coordinator/src/main/java/com/sun/hotspot/igv/coordinator/FolderNode.java:
```diff
@@ -56,15 +61,26 @@ public FolderChildren(Folder folder) {
+        protected void destroyNodes(Node[] nodes) {
+            for (Node n : nodes) {
+                // Each node is only present once in the graphNode map.
+                graphNode.values().remove(n);
```
Is `destroyNodes()` thread-safe here? graphNode is a HashMap instead of ConcurrentHashMap.

navyxliu replied:
This method is by the inner class of RemoveCookie(). I am okay if the framwork guarantees to execute RemoveCookie sequentially.

navyxliu replied:
This method is by the inner class of RemoveCookie(). I am okay if the framwork guarantees to execute RemoveCookie sequentially.

robcasloz replied:
This is guaranteed since `remove()` is called from the event dispatch thread.

navyxliu commented:
hi, @robcasloz, 
This patch works perfectly on MacOS.
--lx

navyxliu commented:
hi, @robcasloz, 
This patch works perfectly on MacOS.
--lx

robcasloz commented:
> hi, @robcasloz, This patch works perfectly on MacOS. --lx

Great, thanks for testing Xin!

I just addressed your comments. Additionally, I realized the changeset did not handle correctly the case where the Outline window is closed and then opened again. This is also fixed in the updated version.

navyxliu reviewed (approved):
LGTM. I am not a review. Still need other reviewers approve it.

chhagedorn reviewed (approved):
Looks good!

robcasloz commented:
> Looks good!

Thanks, Christian!

TobiHartmann reviewed (approved):
Looks reasonable.

robcasloz commented:
> Looks reasonable.

Thanks for reviewing, Tobias!
